packages/core/src/middleware/koa-consent-guard.ts

```diff
@@ -0,0 +1,71 @@
+import { experience } from '@logto/schemas';
+import { type MiddlewareType } from 'koa';
+import { type IRouterParamContext } from 'koa-router';
+import { errors, type Provider } from 'oidc-provider';
+
+import RequestError from '../errors/RequestError/index.js';
+import type Libraries from '../tenants/Libraries.js';
+import type Queries from '../tenants/Queries.js';
+import assertThat from '../utils/assert-that.js';
+
+/**
+ * Guard before allowing auto-consent
+ */
+export default function koaConsentGuard<
+  StateT,
+  ContextT extends IRouterParamContext,
+  ResponseBodyT,
+>(
+  provider: Provider,
+  libraries: Libraries,
+  query: Queries
+): MiddlewareType<StateT, ContextT, ResponseBodyT> {
+  return async (ctx, next) => {
+    const interactionDetails = await provider.interactionDetails(ctx.req, ctx.res);
+    const {
+      params: { token, login_hint: loginHint },
+      session,
+    } = interactionDetails;
```

Comment:
nit: there is a `koa-interaction-details` middleware

Reply:
I saw it, but it seems using it does not reduce complexity that much. Plus, there's no testing util to create context with interaction details context. So I decided not to use it